modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java

```diff
@@ -265,45 +378,116 @@ public MetricRegistry registry(String name) {
         return registries.computeIfAbsent(name, n -> {
             MetricRegistry mreg = new MetricRegistry(name, log);
 
-            notifyListeners(mreg, metricRegCreationLsnrs);
+            notifyListeners(mreg, metricRegCreationLsnrs, log);
 
             return mreg;
         });
     }
 
-    /** {@inheritDoc} */
-    @NotNull @Override public Iterator<MetricRegistry> iterator() {
-        return registries.values().iterator();
+    public <R extends MonitoringRow, D> void list(String name, String description,
+        Class<R> rowClazz, Supplier<ConcurrentMap<?, D>> data, Function<D, R> rowFunc, Consumer<D> clearer) {
```

Comment:
Do we really need `Supplier` for `data` parameter here? Why can't we use data without supplier?

Reply:
A map may be recreated in the class who owns data.

Reply:
In this case, you will have the same problem with the current implementation too (the only difference is you can reenable the list to work correctly). 
But I think it's better to re-register the list if class who owns the data recreates the map. In this case, the list will always be up to date and no reenabling is needed.

Reply:
OK, fixed.